src/D3MHub.sol

```diff
@@ -221,121 +219,122 @@ contract D3MHub {
+        }
+        emit Unwind(ilk, amount);
+    }
 
-        uint256 availableAssets = _pool.maxWithdraw();
+    function _fullUnwind(bytes32 ilk, ID3MPool _pool, address urn) internal {
+        uint256 amount = _pool.maxWithdraw();
+        if (amount > 0) {
+            uint256 toSlip = _min(vat.gem(ilk, urn), amount);
+            require(toSlip <= MAXINT256, "D3MHub/overflow");
 
-        // Unwind amount is limited by how much:
-        // - max reduction desired
-        // - assets available
-        // - dai debt tracked in vat (CDP or free)
-        uint256 amount = _min(
-                            _min(
-                                supplyReduction,
-                                availableAssets
-                            ),
-                            daiDebt
-                        );
-        require(amount <= MAXINT256, "D3MHub/overflow");
-
-        // Determine the amount of fees to bring back
-        uint256 fees = 0;
-        if (assetBalance > daiDebt) {
-            unchecked {
-                fees = assetBalance - daiDebt;
-            }
+            _pool.withdraw(amount);
+            daiJoin.join(address(this), amount);
 
-            if (amount + fees > availableAssets) {
-                unchecked {
-                    fees = availableAssets - amount;
-                }
-            }
+            vat.slip(ilk, urn, -int256(toSlip));
+            vat.move(address(this), vow, amount * RAY);
         }
+        emit Unwind(ilk, amount);
+    }
 
-        if (amount == 0 && fees == 0) {
-            emit Unwind(ilk, 0, 0);
-            return;
+    function _normal(bytes32 ilk, ID3MPool _pool, uint256 Art, uint256 lineWad) internal {
+        uint256 currentAssets = _pool.assetBalance(); // Should return DAI owned by D3MPool
+
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(_pool));
+        uint256 maxWithdraw = _pool.maxWithdraw();
+        if (currentAssets > ink) { // If fees were generated
+            uint256 fixInk = _min(
+                currentAssets - ink, // fees generated
+                ink < lineWad // if previously CDP was under debt ceiling
+                    ? lineWad + maxWithdraw - ink // we can just go above debt ceiling + maxWithdraw
+                    : maxWithdraw // we can just go above current state + maxWithdraw
+            );
+            require(fixInk <= MAXINT256, "D3MHub/overflow");
+            vat.slip(ilk, address(_pool), int256(fixInk)); // Generate extra collateral
+            vat.frob(ilk, address(_pool), address(_pool), address(this), int256(fixInk), 0); // Lock it
+            ink += fixInk;
+            emit Fees(ilk, fixInk);
+        }
+        if (art < ink) { // If there was permissionless DAI paid or fees added as collateral
+            address _vow = vow;
+            uint256 fixArt = ink - art; // Amount of fees + permissionless DAI paid we will now transform to debt
+            delete art;
+            Art += fixArt;
```

Comment:
Not sure if we usually change the values of input variables like this.
Do we have enough place on the stack for another local var?

(this and the `delete art` line make me think if we might want the _fix part as a separate function after all).

Reply:
Ah right, now the `Art` is a param. Yeah in theory we shouldn't be touching this variable. Let me see if I can do something better.